### DISCUSSION TITLE
Ganged Y axis and Auto square

### DISCUSSION THREAD

**Discussion by molove** [phil-barrett/grblHAL-teensy-4.x#31] [Q&A] [answered]
I've finished wiring my control board (for the moment) and have commenced preliminary testing. 

I've flashed Phil's grblHAL_T41E3XGYAS.hex firmware dated 31/10/21. Which if I read correctly means 3 axes, ganged Y axis and auto squaring.

Using IO Sender as the control software, it shows X, Y, Z and A axes. X,Y & Z  all have a red box above them in the GUI but A doesn't, I'm not entirely sure what the red box there signifies. 

![image](https://user-images.githubusercontent.com/20013346/144500330-5950ef77-600b-45cf-ac42-5f55a9221c73.png)

I'm not sure quite how the ganged Y-axis works. So far I have just had one stepper plugged in and have managed set it in motion when plugged into the X,Y,Z and A axes. I assumed that the A axis would be tied to the Y axis, so that when I jogged the Y axis the A axis would move also, but this seems not to be the case as it only moves when I jog the A axis. Am I missing something?

Also, from the Auto square blog post on the grblHAL site, it shows the Y axis having 3 limit switches, 2 used for the auto squaring and one at the other end of the Y axis. Are the limit switches at the other end necessary, or can I just use the 2 for auto square and soft limits for the other end?

Thanks
Piers

**Comment by phil-barrett** [comment]
Not sure why A shows up.  Does the DRO show A also?   In that case, I probably screwed up and put a 4 Axis binary instead of 3 Axis.  What does Grbl/Console show at boot up?   It won't hurt anything.  I'll check it out later today and fix if needed.

Autosquaring happens when homing.  You need a min limit sensor/switch for each Y motor.  You will also need to enable hard limits.  And have your limit targets adjustable so you can tweak them to get your gantry perfectly square. You might want to look at this [blog about autosquaring](https://www.grbl.org/single-post/it-s-hip-to-be-auto-square).  It describes autosquaring in more detail.

**Comment by molove** [comment]
Yes the DRO shows A too

![image](https://user-images.githubusercontent.com/20013346/144503049-dcb69af8-dcbf-4ddb-93cc-0b37b0a91c95.png)

I'm not sure what I'm meant to be looking for in the console.

This is the entirety up to the work offset stuff

GrblHAL 1.1f ['$' or '$HELP' for help]
[VER:1.1f.20210715:]
[OPT:VNMSL2,35,1024,4,0]
[NEWOPT:ENUMS,RT+,ES,TC,ETH]
[FIRMWARE:grblHAL]
[NVS STORAGE:*FLASH]
[DRIVER:iMXRT1062]
[DRIVER VERSION:210725]
[DRIVER OPTIONS:USB.2]
[BOARD:T41U5XBB]
[AUX IO:1,3,0,0]
[IP:0.0.0.0]
[PLUGIN:ENCODER v0.01]

**Comment by phil-barrett** [comment]
Yes, that's the 4 axis binary.  Will fix later today.  Sorry for the confusion.

**Reply by molove** [reply]
No problem, thanks for the swift replies

**Comment by phil-barrett** [comment]
A meeting got cancelled so I fixed it.  zip file is now correct.  Again, sorry for the confusion.

**Reply by molove** [reply]
Thanks I'll download it now

**Comment by molove** [comment]
Phil it looks like there is a problem with the zip archive now, I get an error whilst trying to extract it.

**Reply by phil-barrett** [reply]
I just pulled grblHAL_T41.zip down to my widows 10 machine and was able to copy grblHAL_T41E3XGYAS.hex out with no errors.  Maybe try downloading again?

**Reply by molove** [reply]
I tried downloading multiple times and used 4 different unarchivers on windows (7zip, WinRar, Directory Opus, and Windows Explorer) on 2 laptops and a desktop machine and all failed at extracting the archive. I was able however to extract it when I booted into Linux, but every time I flash grblHAL_T41E3XGYAS it still shows as 4 axes.

**Comment by phil-barrett** [comment]
Huh... Vexed and  Perplexed. Something is not right. 

Here's what I did just now:
1.  Downloaded grblHAL_T41.zip from [here](https://github.com/phil-barrett/grblHAL-teensy-4.x/tree/master/hex_files). 
2. Extracted grblHAL_T41E3XGYAS.hex by going into the zip directory via Windows Explorer and copying it out.
3. Flashed the .hex with teensy.exe. 
4. Ran ioSender which showed it is a 3 Axis build. 

If you are on windows 10, what does the zip file look like in windows file explorer?  A zip directory?  Can you move into it?  Do you have antivirus SW that might be causing a problem? Maybe a program is holding the zip file open and preventing it being written?  Maybe do a full reboot?

**Comment by molove** [comment]
I re-download the zip again and using your method I was able to extract grblHAL_T41E3XGY.hex and grblHAL_T41E3XGYAS.hex, however all of the other files could not be extracted. For example this one:

![zipError](https://user-images.githubusercontent.com/20013346/144596007-7b394ae4-084f-4041-bb8d-5632973f5486.jpg)

So I was able to flash both grblHAL_T41E3XGY.hex and grblHAL_T41E3XGYAS.hex

grblHAL_T41E3XGY.hex comes up as 3 axes with Y controlling the A axis as it should
Console shows:
[OPT:VNMSL,35,1024,3,0] 

![T41E3XGY](https://user-images.githubusercontent.com/20013346/144596844-434c1bf8-73d2-49c7-a204-393f8552c30d.jpg)

grblHAL_T41E3XGYAS.hex however still shows as being 4 axes
Console shows:
[OPT:VNMSL2,999,1024,4,0]

![T41E3XGYAS](https://user-images.githubusercontent.com/20013346/144597084-6e3014bd-e3bd-47ea-bd0e-bbf31aebf420.jpg)

I can't quite figure out what I could have done wrong. I removed all of the old copies, extracted the new download to a different location to previously extracted copies. I flashed the different firmwares to confirm that TeensyLoader was actually flashing the hex files.

Would you mind having another look when you get time. I can use grblHAL_T41E3XGY in the meantime to continue setting up my machine.

Thanks

**Reply by phil-barrett** [reply]
That looks like a windows exception.  Are you fully backed  up? I'd do that ASAP. Also, that is the wrong size for the file in the zip directory (537KB vs 538KB on my development machine here).  I am including the .hex file here as a single zip.
[grblHAL_T41E3XGYAS.zip](https://github.com/phil-barrett/grblHAL-teensy-4.x/files/7651274/grblHAL_T41E3XGYAS.zip)

**Comment by molove** [comment]
I'm pretty sure it isn't a Windows exception, all of my Windows machines barf on the grblHAL_T41.zip from [here](https://github.com/phil-barrett/grblHAL-teensy-4.x/tree/master/hex_files)
Have you tried to extract the whole of the archive yet rather than just copying the single file from it?

![Winrar](https://user-images.githubusercontent.com/20013346/144649086-cc96a333-87d5-47d8-85b7-6897b9cfce78.jpg)
![7zip](https://user-images.githubusercontent.com/20013346/144649097-66bfef6e-f715-4680-b239-ae708bd2e275.jpg)
![Dopus](https://user-images.githubusercontent.com/20013346/144649112-1a78e8b8-3d78-4fec-bf93-e3773066899a.jpg)

I downloaded and extracted grblHAL_T41E3XGYAS.zip from the above post successfully and flashed it but it is still showing as 4 axes. I even tried running IOSender on 2 different laptops with the same results, unfortunately.

Sorry to be such a pain

**Comment by phil-barrett** [comment, marked as answer]
No problem helping you.  This is very odd.  Using windows file explorer in the zip directory, if I highlight all the files and drag them out I get an exception but dragging just one file out I don't.   So, that may be the error you are seeing.  I'll have to do a little research to see if I should use something other than windows for zip files.  I hate when the tools are quirky like this. Thanks for pointing this out.

I went back and rebuilt the 3 Axis Ganged Y, Autosquare hex file. Put it in a zip file, copied it out to a different location, used teensy.exe to load on a board here and verified that it is what it was supposed to be.  I've attached it.  Let me know if that works ok for you. 
[grblHAL_T41E3XGYAS.zip](https://github.com/phil-barrett/grblHAL-teensy-4.x/files/7651874/grblHAL_T41E3XGYAS.zip)

I will go back and reverify grblHAL_T41.zip in the repo. Sorry for the confusion.

**Reply by phil-barrett** [reply]
OK, it looks like the repo zip file WAS corrupted.  Odd that I could pull a single file out but not multiple.  I remade it and tested.  No corruption errors.  Not sure how that happened.  If you could test it out, I would be most appreciative!

Again, thanks for going the distance with me on this. Sorry for suggesting it was your problem. I need to be more willing to consider possibility of the problem being on my end.

**Reply by molove** [reply]
No worries, just got in from a works type xmas party. I'll download now and check it out.
7-Zip is my goto un/archiver -[here]( https://www.7-zip.org/) it's free, open source and will extract most formats.

**Reply by molove** [reply]
I've just flashed the grblHAL_T41E3XGYAS.zip and all is working as it should, 3 axes with Y controlling both Y and A.

I downloaded grblHAL_T41.zip too, and it unarchived with no issues.

Thanks for taking the time to sort this out.

**Comment by AndyCXL** [comment]
May I ask a further question on ganged axis setup for compilation? I intend using X dual-Y Z and A (rotary), and have some defines questions so I know I am pointing in the right direction;

1. I have set "#define Y_GANGED 1" 
2. and "#define N_AXIS 4 // Number of axes" 
3. and "#define ST_ROTATIONAL_MASK (A_AXIS_BIT)"?

Which compiles correctly (TeensyDuino on Mac).

On the BOB, which axis does HAL gang for the second Y? 

- Is it A, in which case I need to define N_AXIS as 5 and declare Rotational for B? ...but this fails to compile with #error "Axis configuration is not supported!"
- Is it A, but still 4 axes, and defining Rotational for B? ...but this fails to compile with #error: 'B_AXIS_BIT' undeclared here (not in a function) #define ST_ROTATIONAL_MASK (B_AXIS_BIT)
- Is it B, having A as the rotational and B as the ganged Y?

The ganged Y axis has to route out to a 'channel' on the BOB, but which axis maps to which channel is unclear, and the compilation errors being reported don't (for me) clarify things as yet. I want to be able to compile, rather than use pre-built .hex files, as there are various other issues I want to explore before transplanting the current controller in my CNC (ShapeOKO with added Rotary, KeyeStudio 2560-based Re-Rap BOB, Grbl Mega-5X) such as limit switch polarities (I have active limit switches, which I believe are NO), Emergency Stop polarities (what I have today is NO wired to feed-hold rather than stop), PWM signal modes for my Huangyang VFD (currently wired as 0-5V PWM not RS485) etc, etc.

So back to the original question - what is the GRBL HAL logic for ganging Y? Any assistance gratefully received

**Reply by phil-barrett** [reply]
Y gangs to B in your case because you have defined 4 axes in config.h. The rotary axis is A.

The logic for ganging is "next available motor", in alphabetical order.  In the case of N_AXIS is 3, A is the next available motor so Y is ganged to A. In the case of N_AXIS is 4, Y is ganged to B since it is the next available motor.  If you define N_AXIS 5, you will get an error because there is no next available motor.

Hope that is clear.
Phil

**Reply by AndyCXL** [reply]
Phil - thank you for coming back so promptly. I've just been probing the Stepper signals and had concluded that, with 4 axes declared, the next free axis is the Y gang - ie: B.
Very impressed with your BOB, having received mine just this weekend. Now I have the axis ordering resolved, I can move to a 'desktop' setup to establish active limit switch NO or NC behaviour, etc., and experiment with Trinamic drivers - is there a 'hobby end' preference for these drivers? I've used TMC2209 step-sticks before, but have no familiarity beyond those

**Reply by phil-barrett** [reply]
Driver selection depends on your motor size. Dm542 is good for mid size motors.  Make sure you get ones that can handle 125% of your motor current.

On limit and input switch, grblHAL defaults to NC which will cause some initial confusion. The First Run Settings section of the grblHAL wiki will help you through that.

**Reply by AndyCXL** [reply]
Got this up and running on a test rig, and noted the input 'polarity' defaults. Have opted for DM3230 drivers as 3A gives good headroom for my NEMA-23 steppers (Std fitment on ShapeOKO) which are rated to 2.5A abs max and DM3230 is physically smaller than the 542 which helps with packaging the controller. Exploring with GSender as the non-rotary capable sender, and UGS for rotary, with GRBL HAL set into compatibility 1, I am MAC and RPI based, so IOSender is awkward unless/until there is a cross-platform version.
Making progress! Happy days!!
